File: CS/GroupFootersWPF/GroupFooter.cs
using DevExpress.Xpf.Data;
using DevExpress.Xpf.Grid;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;
using System.Windows.Controls;

namespace GroupFootersWPF
{
    public class GroupFooter : Control
    {
        public static readonly DependencyProperty GroupRowDataProperty =
            DependencyProperty.Register("GroupRowData", typeof(MyGroupRowData), typeof(GroupFooter), new PropertyMetadata(null));
        public static readonly DependencyProperty RowHandleProperty =
            DependencyProperty.Register("RowHandle", typeof(RowHandle), typeof(GroupFooter), new PropertyMetadata(new RowHandle(GridControl.InvalidRowHandle), (d, e) => ((GroupFooter)d).RefreshContent()));

        public GroupFooter()
        {
            DefaultStyleKey = typeof(GroupFooter);
        }

        public MyGroupRowData GroupRowData
        {
            get { return (MyGroupRowData)GetValue(GroupRowDataProperty); }
            set { SetValue(GroupRowDataProperty, value); }
        }
        public RowHandle RowHandle
        {
            get { return (RowHandle)GetValue(RowHandleProperty); }
            set { SetValue(RowHandleProperty, value); }
        }

        public MyTableView View { get { return DataContext == null ? null : (MyTableView)((RowData)DataContext).View; } }

        public override void OnApplyTemplate()
        {
            base.OnApplyTemplate();
            RefreshContent();
        }
        public bool IsLastRowInGroup(int groupRowHandle)
        {
            GridControl gridControl = View.DataControl as GridControl;
            if (!gridControl.IsValidRowHandle(groupRowHandle))
                return false;
            int childrenCount = gridControl.GetChildRowCount(groupRowHandle);
            int lastRowHandle = gridControl.GetChildRowHandle(groupRowHandle, childrenCount - 1);
            return gridControl.GetRowVisibleIndexByHandle(lastRowHandle) <= gridControl.GetRowVisibleIndexByHandle(RowHandle.Value);
        }
        public void RefreshContent()
        {
            if (View == null)
                return;
            RowHandle groupRowHandle = GetParentRowHandle();
            View.AddToCacheGroupFooter(this);
            UpdateVisibility(groupRowHandle);
            if (Visibility == Visibility.Collapsed)
                return;
            UpdateGroupFooterSummaryContent(groupRowHandle);
            UpdateOffset();
        }
        void UpdateOffset()
        {
            double offset = View.LeftGroupAreaIndent * ((GridControl)View.DataControl).GroupCount;
            Margin = new Thickness(-offset, 0, 0, 0);
        }
        void UpdateGroupFooterSummaryContent(RowHandle groupRowHandle)
        {
            if (GroupRowData == null)
                GroupRowData = View.CreateGroupRowData(groupRowHandle);
            else
                View.UpdateGroupRowData(GroupRowData, groupRowHandle);
        }
        RowHandle GetParentRowHandle()
        {
            return new RowHandle(((GridControl)View.DataControl).GetParentRowHandle(RowHandle.Value));
        }
        void UpdateVisibility(RowHandle groupRowHandle)
        {
            Visibility = IsLastRowInGroup(groupRowHandle.Value) ? Visibility.Visible : Visibility.Collapsed;
        }
    }
}

File: VB/GroupFootersWPF/GroupFooter.vb
Imports DevExpress.Xpf.Data
Imports DevExpress.Xpf.Grid
Imports System.Windows
Imports System.Windows.Controls

Namespace GroupFootersWPF

    Public Class GroupFooter
        Inherits Control

        Public Shared ReadOnly GroupRowDataProperty As DependencyProperty = DependencyProperty.Register("GroupRowData", GetType(MyGroupRowData), GetType(GroupFooter), New PropertyMetadata(CType(Nothing, PropertyChangedCallback)))

        Public Shared ReadOnly RowHandleProperty As DependencyProperty = DependencyProperty.Register("RowHandle", GetType(RowHandle), GetType(GroupFooter), New PropertyMetadata(New RowHandle(DataControlBase.InvalidRowHandle), Sub(d, e) CType(d, GroupFooter).RefreshContent()))

        Public Sub New()
            DefaultStyleKey = GetType(GroupFooter)
        End Sub

        Public Property GroupRowData As MyGroupRowData
            Get
                Return CType(GetValue(GroupRowDataProperty), MyGroupRowData)
            End Get

            Set(ByVal value As MyGroupRowData)
                SetValue(GroupRowDataProperty, value)
            End Set
        End Property

        Public Property RowHandle As RowHandle
            Get
                Return CType(GetValue(RowHandleProperty), RowHandle)
            End Get

            Set(ByVal value As RowHandle)
                SetValue(RowHandleProperty, value)
            End Set
        End Property

        Public ReadOnly Property View As MyTableView
            Get
                Return If(DataContext Is Nothing, Nothing, CType(CType(DataContext, RowData).View, MyTableView))
            End Get
        End Property

        Public Overrides Sub OnApplyTemplate()
            MyBase.OnApplyTemplate()
            RefreshContent()
        End Sub

        Public Function IsLastRowInGroup(ByVal groupRowHandle As Integer) As Boolean
            Dim gridControl As GridControl = TryCast(View.DataControl, GridControl)
            If Not gridControl.IsValidRowHandle(groupRowHandle) Then Return False
            Dim childrenCount As Integer = gridControl.GetChildRowCount(groupRowHandle)
            Dim lastRowHandle As Integer = gridControl.GetChildRowHandle(groupRowHandle, childrenCount - 1)
            Return gridControl.GetRowVisibleIndexByHandle(lastRowHandle) <= gridControl.GetRowVisibleIndexByHandle(RowHandle.Value)
        End Function

        Public Sub RefreshContent()
            If View Is Nothing Then Return
            Dim groupRowHandle As RowHandle = GetParentRowHandle()
            View.AddToCacheGroupFooter(Me)
            UpdateVisibility(groupRowHandle)
            If Visibility = Visibility.Collapsed Then Return
            UpdateGroupFooterSummaryContent(groupRowHandle)
            UpdateOffset()
        End Sub

        Private Sub UpdateOffset()
            Dim offset As Double = View.LeftGroupAreaIndent * CType(View.DataControl, GridControl).GroupCount
            Margin = New Thickness(-offset, 0, 0, 0)
        End Sub

        Private Sub UpdateGroupFooterSummaryContent(ByVal groupRowHandle As RowHandle)
            If GroupRowData Is Nothing Then
                GroupRowData = View.CreateGroupRowData(groupRowHandle)
            Else
                View.UpdateGroupRowData(GroupRowData, groupRowHandle)
            End If
        End Sub

        Private Function GetParentRowHandle() As RowHandle
            Return New RowHandle(CType(View.DataControl, GridControl).GetParentRowHandle(RowHandle.Value))
        End Function

        Private Sub UpdateVisibility(ByVal groupRowHandle As RowHandle)
            Visibility = If(IsLastRowInGroup(groupRowHandle.Value), Visibility.Visible, Visibility.Collapsed)
        End Sub
    End Class
End Namespace
